### PR TITLE
kustomize: Add support for OCI based helm repos

### LINKED ISSUE #4381: `helmCharts` repo field doesn't work with OCI registry
**Describe the bug**

Kustomize fails to pull charts through `helmCharts` yaml that are hosted within an OCI registry.

** Example **

Create a `kustomize.yaml` file:

```
helmCharts:
  - name: someName
    includeCRDs: true
    valuesFile: values.yaml
    releaseName: myRelease
    version: 0.0.26
    repo: oci://myRegistry
```

Run the command `kustomize build --enable-helm --helm-command $helmCommand ./folder`

Kustomize will fail to pull the helm chart and produce the following error:

```
Error: Error: looks like "oci://myRegistry" is not a valid chart repository or cannot be reached: object required
: unable to run: '/usr/sbin/helm pull --untar --untardir <someDir> --repo oci://myRegistry myRelease --version 0.0.26' with env=[HELM_CONFIG_HOME=/tmp/kustomize-helm-991395685/helm HELM_CACHE_HOME=/tmp/kustomize-helm-991395685/helm/.cache HELM_DATA_HOME=/tmp/kustomize-helm-991395685/helm/.data] (is '/usr/sbin/helm' installed?)
```

Running a regular `helm pull` works fine on the registry, but pulling with `--repo` causes issues:
```
# Working
$ helm pull oci://myRegistry/myRelease --version 0.0.26
Pulled: myRegistry/myRelease:0.0.26
Digest: sha256:14e1bc1ec1d0147eb8b41f081faae2ecea63e862ad4259d3a9cd1c7d1584be63

# Not working with `--repo`
$ helm pull myRelease --version 0.0.26 --repo myRegistry
Error: could not find protocol handler for: 

$ helm pull myRelease --version 0.0.26 --repo oci://myRegistry
Error: looks like "oci://myRegistry" is not a valid chart repository or cannot be reached: object required
```

I also cannot work around this by just adding it as a repo:
```
$ helm repo add reg myRegistry
Error: could not find protocol handler for: 

$ helm repo add reg oci://myRegistry
Error: looks like "oci://myRegistry" is not a valid chart repository or cannot be reached: object required
```

**Expected output**

Expected Kustomize to be able to pull the chart and apply the `values.yaml` to the chart. The solution would be for `oci://` urls, to not use the `--repo` flag when performing a helm pull.

**Kustomize version**

$ kustomize version
{Version:kustomize/v4.3.0 GitCommit:cd17338759ef64c14307991fd25d52259697f1fb BuildDate:2021-08-24T19:24:28Z GoOs:linux GoArch:amd64}

$ helm version
version.BuildInfo{Version:"v3.7.2", GitCommit:"663a896f4a815053445eec4153677ddc24a0a361", GitTreeState:"clean", GoVersion:"go1.16.10"}

**Platform**

Ubuntu

### PR DESCRIPTION
**Pull request by jkroepke**
[Bitnami recently change from classical helm repositories to OCI based repositories](https://blog.bitnami.com/2023/04/httpsblog.bitnami.com202304bitnami-helm-charts-now-oci.html), which may have an impact of the support for helm charts from OCI repositories.

this PR add support for fetch charts from OCI repositories.

Example:

```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
helmCharts:
- name: external-dns
#  repo: https://charts.bitnami.com/bitnami
  repo: oci://registry-1.docker.io/bitnamicharts
  version: 6.19.2
  releaseName: external-dns
  namespace: external-dns
  valuesFile: values.yaml
```

This PR does not implement a breaking change.

Fixes #4381
ref #4614

### CODE REVIEW

File: api/internal/builtins/HelmChartInflationGenerator.go
Comment (jkroepke) [reply api/internal/builtins/HelmChartInflationGenerator.go:289]: You are right.

```suggestion
			args = append(args, "--repo", p.Repo, p.Name) 
```

File: plugin/builtin/helmchartinflationgenerator/HelmChartInflationGenerator.go
Comment (jkroepke) [inline plugin/builtin/helmchartinflationgenerator/HelmChartInflationGenerator.go:295]:
```suggestion
			args = append(args, "--repo", p.Repo, p.Name) 
```

**Review by soudaburger — APPROVED** [review]

**Comment by koba1t** [comment]
Hi @jkroepke
Thanks for your contribution!

This PR almost looks good to me.
So, it looks like the lint failed on CI. Could you check it and fix it?

**Comment by DanielYWoo** [comment]
Do we have an ETA for this to be released?

**Comment by koba1t** [comment]
@jkroepke 

I found the below error when running on my laptop. Could you fix it?

```
api/krusty/helmchartinflationgenerator_test.go:49:7: var-naming: const expectedHelmExternalDns should be expectedHelmExternalDNS (revive)
const expectedHelmExternalDns = `
      ^
make[1]: *** [lint] Error 1
make: *** [lint] Error 2
```

**Comment by jkroepke** [comment]
Done

**Comment by koba1t** [comment]
@jkroepke

Thanks!
Almost looks good to me!
/lgtm

@natasha41575 
Could you check this PR and rerun GitHub Actions?

**Comment by koba1t** [comment]
/cc @natasha41575 
Could you check this PR and rerun GitHub Actions?

**Comment by robertsa1-els** [comment]
Any movement on this.  It is not long till Bitnami are due to drop support and GitHub Scalesets is OCI only.
Thanks in advance 👍

**Comment by jkroepke** [comment]
@koba1t how to proceed here? I'm bored of constantly rebase this PR.

**Comment by koba1t** [comment]
@jkroepke 
We need to wait for the review of the maintainers.

/cc @natasha41575 @annasong20

**Comment by jkroepke** [comment]
@koba1t because of the rebase, the lgtm flag has been removed

**Comment by koba1t** [comment]
Thanks
/lgtm

**Comment by jkroepke** [comment]
@koba1t are you able to trigger to CI here? Just to confirm that everything is green.

**Comment by koba1t** [comment]
@jkroepke
> @koba1t are you able to trigger to CI here?

Sorry... I don't have that permission. Only the maintainer can do that operation...

**Comment by jkroepke** [comment]
Alright, I hope that  @natasha41575 @annasong20 will find some time, soon.

**Comment by tolgaio** [comment]
just pinging @natasha41575 @annasong20 again, we've been waiting for this for a long time now as well, it feels we are nearly there, could you please trigger the CI and help us here to get this merged. thanks a bunch!

**Review by FrenchBen — APPROVED** [review]
LGTM

**Comment by diegoluisi** [comment]
@natasha41575

**Comment by diegoluisi** [comment]
Hope that @annasong20, @yuwenma or @natasha41575 approve this PR soon..

**Comment by shakefu** [comment]
3 months into merging a 140 line PR? Is this dead in the water?

**Comment by jkroepke** [comment]
@natasha41575 do you have any idea to proceed here? @KnVerey seems unresponsive and I see a lot of recent contribution from you on main.

/assign @natasha41575

**Comment by spstarr** [comment]
Really needing this feature with Config Sync

**Review by cheelim1 — APPROVED** [review]
LGTM

**Comment by natasha41575** [comment]
I am not sure why the tests aren't running, the button to run them isn't there. Could you try pushing something to see if it runs the tests?

**Comment by jkroepke** [comment]
@natasha41575  now it says "workflows awaiting approval"

**Comment by hawkesn** [comment]
The tests look like they passed? I think it just needs the lgtm labels?
image

**Comment by soudaburger** [comment]
LGTM

**Comment by soudaburger** [comment]
Can we get the LGTM and any other necessary labels on this.  We're so close finally!

**Comment by bruceoutdoors** [comment]
LGTM

**Comment by stesme** [comment]
Can someone put an LGTM on this PR, please 🙏

**Review by bruceoutdoors — APPROVED** [review]

**Comment by joshuabaird** [comment]
@annasong20 Is there anything we can do to help move this forward and get it merged?

**Comment by tolgaio** [comment]
🍿

**Comment by NigelVanHattum** [comment]
@natasha41575 Could you take a look at the PR again. All tests are green, we only need the LGTM tag.

**Comment by pkruk** [comment]
ping?

**Comment by ramizpolic** [comment]
Any info on when we can expect this PR to be merged?

**Review by annasong20 — CHANGES_REQUESTED** [review]
@jkroepke I think this PR mostly looks good! My concerns are relatively minor. Would value @natasha41575's feedback, as she has more Helm expertise, but otherwise I can give my approval after the requested changes are fixed.

I sincerely apologize for the huge delay in my review. It's a continual combination of limited bandwidth and a long backlog.

File: plugin/builtin/helmchartinflationgenerator/HelmChartInflationGenerator.go
Comment (jkroepke) [reply plugin/builtin/helmchartinflationgenerator/HelmChartInflationGenerator.go:293]: At least, there is one https://github.com/kubernetes-sigs/kustomize/pull/5167/files#diff-28bb15d994f41f933c7d9304d77227b0805245111b7590b293456733b8758292R188 or I'm wrong?

**Comment by jkroepke** [comment]
@annasong20 Thanks for your review, I applied all your suggestion. They make sense my eyes.

**Comment by jmrr** [comment]
Thank goodness I found this and the #4381 issue, I was already going to beat my head against the wall, as the [charts](https://github.com/actions/actions-runner-controller/tree/master/charts) we're using are only published in an OCI repo. Subscribed to the PR and looking forward to seeing it released. Thank you!

File: plugin/builtin/helmchartinflationgenerator/HelmChartInflationGenerator.go
Comment (koba1t) [reply plugin/builtin/helmchartinflationgenerator/HelmChartInflationGenerator.go]: I think @annasong20 wants to add two type tests one of an OCI repo doesn't end with a "/" and the OCI repo ends with a "/" word.
Could you add another type of test?

**Comment by koba1t** [comment]
Hi @jkroepke

I'll review this PR instead of @annasong20 because she is still busy.
I was promoted to `approver` last month, and that means I can `approve` this PR.

Looks like almost good for your PR.
I have two nit comments. Could you check these?
I'll merge this PR after these are fixed.

File: go.work.sum
Comment (koba1t) [inline go.work.sum:1]: Could you execute the `go work sync` command?
I found the diff after exec this command in my laptop.

Comment (jkroepke) [reply go.work.sum:1]: I run `go work sync` which introduces a lot of changes in a go.mod/go.sum files.

In case, that not intended, please commit to this PR directly, since "Allow edits and access to secrets by maintainers" is enable and I can not validate whats is correct here.

Comment (koba1t) [reply go.work.sum:1]: > In case, that not intended,

No, You completed what I wanted!

File: plugin/builtin/helmchartinflationgenerator/HelmChartInflationGenerator.go
Comment (koba1t) [reply plugin/builtin/helmchartinflationgenerator/HelmChartInflationGenerator.go]: @jkroepke 
Could you check it?

**Comment by DanielYWoo** [comment]
appreciate your work but I think the change process to k8s is an overkill.

**Comment by NigelVanHattum** [comment]
@koba1t Can we get this merged? :-)

**Comment by jkroepke** [comment]
did get the notify for that, I will take not of it!

**Comment by jkroepke** [comment]
@koba1t I added a test and do a merge from master

**Comment by koba1t** [comment]
@jkroepke

Thank you for adding another test and for your continued attention to this PR over an extended period. 
This PR looks good to me. I'll merge this PR!

/approve

**Comment by aristosvo** [comment]
Thanks @koba1t and @jkroepke!

Can we expect a release with this functionality in the near future?

**Comment by natasha41575** [comment]
This is available in the latest release.

**Comment by MPV** [comment]
Beautiful!

Looking to add it into ArgoCD over at:
- https://github.com/argoproj/argo-cd/pull/16054#issuecomment-1772815169

**Comment by zimmertr** [comment]
Is this not working with GHCR? The [repo](ghcr.io/sergelogvinov/charts/proxmox-csi-plugin) is public?

```yaml
helmCharts:
  - name: csi-proxmox
    repo: oci://ghcr.io/sergelogvinov/charts/proxmox-csi-plugin
    version: 0.1.15 # https://ghcr.io/sergelogvinov/charts/proxmox-csi-plugin
    releaseName: csi-proxmox
    namespace: csi-proxmox
```

> Error: failed to authorize: failed to fetch anonymous token: unexpected status from GET request to https://ghcr.io/token?scope=repository%!A(MISSING)sergelogvinov%!F(MISSING)cha
rts%!F(MISSING)proxmox-csi-plugin%!F(MISSING)csi-proxmox%!A(MISSING)pull&service=ghcr.io: 403 Forbidden

**Comment by koba1t** [comment]
@zimmertr 

We don't plan to support `private repository or registry authentication` in `helmCharts`.
https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/helmcharts/#the-current-builtin

**Comment by koba1t** [comment]
@zimmertr

I think you made a mistake in your `kustomization.yaml` file; the below file works on my computer.

```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

helmCharts:
  - name: proxmox-csi-plugin
    repo: oci://ghcr.io/sergelogvinov/charts
    version: 0.1.15 # https://ghcr.io/sergelogvinov/charts/proxmox-csi-plugin
    releaseName: csi-proxmox
    namespace: csi-proxmox
```

https://kubectl.docs.kubernetes.io/references/kustomize/builtins/#_helmchartinflationgenerator_

**Comment by zimmertr** [comment]
Doi, Thank you :)

**Comment by marcusnh** [comment]
Hello, I have the same problem with [my setup](https://github.com/argoproj/argo-cd/issues/16894) to Azure Container Registery:
```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
# Referencing a public repo outside the main repository

helmCharts:
  - name: <ARO-REPO-CHART-NAME>
    repo: oci://ACR-NAME.azurecr.io/
    version: 0.1.6-5
    releaseName:ARO-REPO-NAME
    namespace: poseidon2-dev
    valuesFile: values.yaml
```
The error message is also similar:
```
level=error msg="finished unary call with code Unknown" error="Manifest generation error (cached): `kustomize build <path to cached source>/applicationsets/dev/demo-helm-2 --enable-helm` failed exit status 1: Error: Error: failed to authorize: failed to fetch anonymous token: unexpected status from GET request to https://<ACR-NAME>.azurecr.io/oauth2/token?scope=repository%!A(MISSING)<ARO-REPO-NAME>%!F(MISSING)<ARO-REPO-NAME>%!A(MISSING)pull&service=<ACR-NAME>.azurecr.io: 401 Unauthorized\n: unable to run: 'helm pull --untar --untardir <path to cached source>/applicationsets/dev/demo-helm-2/charts oci://<ACR-NAME>.azurecr.io/<ARO-REPO-NAME>/<ARO-REPO-NAME> --version 0.1.6-5' with env=[HELM_CONFIG_HOME=/tmp/kustomize-helm-3509023410/helm HELM_CACHE_HOME=/tmp/kustomize-helm-3509023410/helm/.cache HELM_DATA_HOME=/tmp/kustomize-helm-3509023410/helm/.data] (is 'helm' installed?): exit status 1" grpc.code=Unknown grpc.method=GenerateManifest grpc.service=repository.RepoServerService grpc.start_time="2024-01-17T09:44:38Z" grpc.time_ms=2.234 span.kind=server system=grpc
```

**Comment by koba1t** [comment]
According to your error message, you failed to exec `helm pull` command.
Please try to exec `helm pull` directly to check if the helm chart is valid.

**Comment by marcusnh** [comment]
This works fine on my local computer. I'm able to run helm pull:
```
helm pull oci://ACR-NAME.azurecr.io/ARO-REPO-CHART-NAME  --version 0.1.6-5
Pulled: ACR-NAME.azurecr.io/ARO-REPO-CHART-NAME:0.1.6-5
Digest: sha256:9be2d3dd16cee312918f2b321453683e8b-NOT-THE-SAME-57b
```
I have also tested the chart with a ArgoCD application and that works fine with the kubernetes secret passed to my ArgocD instance:
```
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: test-<ARO-REPO-NAME>
  namespace: gitops-developers # argocd instance namespace
spec:
  source:
    chart: <ARO-REPO-CHART-NAME>
    repoURL: <ARO-NAME>.azurecr.io
    targetRevision: 0.1.6-5
    helm:
      values: |
        application_name: "<ARO-REPO--CHART-NAME>-test"
        namespace: poseidon2-test

  destination:
    namespace: poseidon2-test
    server: https://kubernetes.default.svc
  project: poseidon2
  syncPolicy:
    automated:
      prune: true
      selfHeal: true
      allowEmpty: true
    # syncOptions:
    #   - Replace=true
```

**Comment by koba1t** [comment]
> I have also tested the chart with a ArgoCD application and that works fine with the kubernetes secret passed to my ArgocD instance:

I apologize for not knowing the details of ArgoCD. Do you use any credentials to pull helm charts?
kustomize doesn't support private repository or registry authentication for helm pull.
https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/helmcharts/

**Comment by TeamPoseidonOCP** [comment]
To get access to the private ACR, I need some form of authentication, and creating helm credentials with a Kubernetes secret has been sufficient to get access with ARgoCD. 

In the link you sent, it says: "We will not add support for:
* private repository or registry authentication
* OCI registries
* other large features that increase the complexity of the feature and/or have significant security implications."

OCI registries are now supported. Are there no planned features for enabling private repositories or registry authentication? This would help us a lot and simplify the setup for our developers.

**Comment by koba1t** [comment]
Sorry, we don't have a plan to add more functions to helmCharts now.

**Comment by MrFreezeex** [comment]
> To get access to the private ACR, I need some form of authentication, and creating helm credentials with a Kubernetes secret has been sufficient to get access with ARgoCD.

FYI I put some observation possible followup on this (private oci support): https://github.com/kubernetes-sigs/kustomize/issues/5407#issuecomment-1804441770 and there is two possible (ugly) workaround that you could do with existing versions

**Comment by felixlut** [comment]
I think the above issues are related to this ticket: https://github.com/kubernetes-sigs/kustomize/issues/5407

My workaround for the time being is to run kustomize build in CI (with a short bash script as a workaround to auth to my private helm registry), generate a manifest.lock.yaml containing the entire k8s manifest, and have Argo cd look at that directly. Essentially move the kustomize build step from ArgoCD to CI instead, since I have more control there.

This is also consistent with the rendered manifests pattern:
https://akuity.io/blog/the-rendered-manifests-pattern

With all that said, I really think that kustomize should respect the local Helm credentials, as mentioned in the issue I linked above

**Comment by marcusnh** [comment]
Thank you for the feedback. I saw the workarounds, and there is the possibility to add them with a config map to the ArgoCD instance, but I think we will have to use another approach until this is supported. Using the workaround provided in the ticket [#5407](https://github.com/kubernetes-sigs/kustomize/issues/5407) will fix the problem, but it is not the cloud-native way that we prefer to do things. I think we will have to stick with ArgoCD applications and helm integration until this is supported.  
Is there a feature request that I can follow or contribute to?

**Comment by jkroepke** [comment]
About the authentication issues, you can use helm login, crane login or docker login before running kustomize. The helm OCI integration will re-use the docker credentials for authentication. If you are running private ACR with Azure RBAC, consider to use `az acr login` instead `docker login`.

**Comment by jkroepke** [comment]
Tested with ghcr.io. Works fine.

**Comment by MrFreezeex** [comment]
Actually it does seems to work with helm 3.13.3 indeed thanks! Pretty sure there is something that affected the usage of helm through kustomize that got fixed at the helm level in the last couple of months as I was not alone affected by this so it would be weird that we all messed up at the same time 🤔. Anyway looks fixed to me!

**Comment by marcusnh** [comment]
Is there an update from v3.13.2 to v3.13.3 that has fixed the authentication issues?
Our infrastructure is using ArgoCD to provision applications for our developers, so we need a way of doing the `helm registry login` command through that API.  Similar to ArgoCD already does today through ArgoCD applications.
>  If you are running private ACR with Azure RBAC, consider to use az acr login instead docker login.
Need to be able to use any registry,  not just ARC.

**Comment by marcusnh** [comment]
It is kustomize that doesn't support private OCI registries, so the issue needs to be fixed there and not in ArgoCD. This feature request is only to make kustomize more compatible with ArgoCD and Helm. It is possible to add a workaround, but it is not very cloud-native. Take a look at this issue; might clear up why this is a kustomize problem and not an ArgoCD error:
- https://github.com/argoproj/argo-cd/issues/16623

**Comment by jkroepke** [comment]
It does support it. See above.

ArgoCD is should provide the authenticated context. Argo does this for helm, too.

You can also run helm registry login before running kustomize.

**Pull request merged**
